### ISSUE TITLE
iperf 3.1.6 UDP block size exceeds TCP MSS 0

### ISSUE THREAD

**Issue by fl1212** [esnet/iperf#604]
Iperf 3.1.6
Win10 TH2

I am running iperf to generate packets between my host PC to a VM.
Following command works fine but I receive a warning and I do not know exactly if I can ignore it or not:
```
iperf3.exe -c 10.10.10.2 -t 60 -O 1 -fm  -V  -u  -b 1000m  -R  -l 5000.

Warning:  UDP block size 5000 exceeds TCP MSS 0, may result in fragmentation / drops
Summary Results:
[ ID] Interval           Transfer     Bandwidth       Jitter    Lost/Total Datagrams
[  4]   0.00-60.00  sec  6.90 GBytes   987 Mbits/sec  148.369 ms  153028/1486702 (10%)  
[  4] Sent 1486702 datagrams
[SUM]  0.0-60.0 sec  35886 datagrams received out-of-order
CPU Utilization: local/receiver 27.2% (9.4%u/17.9%s), remote/sender 41.4% (0.9%u/40.5%s)
```
I also tried without the -l option and I receive the same warning but the Bandwidth is 548Mbit/sec and the warning is:  UDP block size 1460 exceeds TCP MSS 0, may result in fragmentation.

EDIT:  Reformatted for command output legibility. --bmah

**Comment by bmah888** [comment]
Based on this output it looks like iperf3 wasn't successful in getting the MSS for the control connection.  You can probably ignore this message, it's just a warning.  Note that iperf3 doesn't officially support Windows, but it's possible this isn't the only platform where this can happen.

I suppose we can do a check to make sure that we actually have a valid (> 0) MSS before printing that message.

**Comment by DennisEdlund** [comment]
UDP have datagrams and not TCP segments so it should definitely return with
an error.
MSS parameter should not even be used on UDP.
Maximum size of UDP datagram without header inside IP datagram  = 65527 -
20 bytes ( ip header) = 65507 bytes
The default TCP Maximum Segment Size is 536.
But still, fragmentation of data packets happens above MTU size.

^^^^^^^^^^^^^^^^^^^^

Dennis Edlund
+46 (0) 707-484571
dennis@detek.se


DeTeK AB
Roxnäsvägen 27
644 36 TORSHÄLLA
SWEDEN
www.detek.se


^^^^^^^^^^^^^^^^^^^^

2017-06-20 20:40 GMT+02:00 Bruce A. Mah <notifications@github.com>:

**Comment by bmah888** [comment]
@DennisEdlund :  iperf3 uses the MSS of the control connection (which runs over TCP) to make an estimate of the maximum payload size for packets in a UDP test, in order to avoid fragmentation.  The assumption is that TCP and UDP traffic will run over the same path (and therefore have the same PMTU), and we try to compensate for the different header sizes in the code.  The message that is printed is only a warning, and it does not affect the actual test behavior of iperf3 in any way.  The user still free to set a maximum send size with `-l` that is larger than this.

**Comment by DennisEdlund** [comment]
Okey, MSS clamping.
That makes sense, smart usage.
Have PMTUD been tried and not working well?
Another pratical way would be as in PMTUD do test sends with "Don't Fragment flag" (DF) and starting with a large packet from a list of common values until you have packet going through and then do a manual search to find the correct MTU.

Could there be a parameter option to "Don't Fragment" (DF)  on transmission?

This is out of the thread but have you tried an auto send option?
Where it will send blocks of packages with a timed interval between each (like -b xx) and wait for a receive OK and when it has a recieved OK it will send again with a shorter/larger interval.
This is to automatically find the correct threshold for the bandwidth?

**Comment by bmah888** [comment]
Re-open as a reminder to merge to 3.2.x

**Comment by bmah888** [comment]
Merged to 3.2 in 6663be4144873b2751e7ae48b0161663ecf78d00 and 58332f8154e2140e40a6e0ea060a418138291718.  Closing.
